src/py/flwr/client/secure_aggregation/secaggplus_handler.py

```diff
@@ -0,0 +1,453 @@
+
+from .handler import SecureAggregationHandler
+
+STAGE_SETUP = "setup"
+STAGE_SHARE_KEYS = "share_keys"
+STAGE_COLLECT_MASKED_INPUT = "collect_masked_input"
+STAGE_UNMASKING = "unmasking"
+STAGES = (STAGE_SETUP, STAGE_SHARE_KEYS, STAGE_COLLECT_MASKED_INPUT, STAGE_UNMASKING)
+
+
+@dataclass
+# pylint: disable-next=too-many-instance-attributes
+class SecAggPlusState:
+    """State of the SecAgg+ protocol."""
+
+    sid: int = 0
+    sample_num: int = 0
+    share_num: int = 0
+    threshold: int = 0
+    clipping_range: float = 0.0
+    target_range: int = 0
+    mod_range: int = 0
+
+    # sk, pk stand for secret key, public key
+    sk1: bytes = b""
+    pk1: bytes = b""
+    sk2: bytes = b""
+    pk2: bytes = b""
+    # random seed for generating the private mask
+    rd_seed: bytes = b""
+
+    rd_seed_share_dict: Dict[int, bytes] = field(default_factory=dict)
+    sk1_share_dict: Dict[int, bytes] = field(default_factory=dict)
+    # the dict of the shared secrets from sk2
+    ss2_dict: Dict[int, bytes] = field(default_factory=dict)
+    public_keys_dict: Dict[int, Tuple[bytes, bytes]] = field(default_factory=dict)
+
+    client: Optional[Union[Client, NumPyClient]] = None
+
+
+class SecAggPlusHandler(SecureAggregationHandler):
+    """Message handler for the SecAgg+ protocol."""
+
+    _shared_state = SecAggPlusState()
+    _current_stage = STAGE_UNMASKING
+
+    def handle_secure_aggregation(
+        self, named_values: Dict[str, Value]
+    ) -> Dict[str, Value]:
+        """Handle incoming message and return results, following the SecAgg+ protocol.
+
+        Parameters
+        ----------
+        named_values : Dict[str, Value]
+            The named values retrieved from the SecureAggregation sub-message
+            of Task message in the server's TaskIns.
+
+        Returns
+        -------
+        Dict[str, Value]
+            The final/intermediate results of the SecAgg+ protocol.
+        """
+        if not isinstance(self, (Client, NumPyClient)):
+            raise TypeError(
+                "The subclass of SecAggPlusHandler must be "
+                "the subclass of Client or NumPyClient."
+            )
+
+        self._check_stage(named_values)
+        stage = str(named_values.pop("stage"))
+        self._current_stage = stage
+
+        self._check_named_values(named_values)
+        if stage == STAGE_SETUP:
+            self._shared_state = SecAggPlusState(client=self)
+            return _setup(self._shared_state, named_values)
+        if stage == STAGE_SHARE_KEYS:
+            return _share_keys(self._shared_state, named_values)
+        if stage == STAGE_COLLECT_MASKED_INPUT:
+            return _collect_masked_input(self._shared_state, named_values)
+        if stage == STAGE_UNMASKING:
+            return _unmasking(self._shared_state, named_values)
+        raise ValueError(f"Unknown secagg stage: {stage}")
+
+    def _check_stage(self, named_values: Dict[str, Value]) -> None:
```

Comment:
This could be a function:
```suggestion
    def _check_stage(self, current_stage: str, named_values: Dict[str, Value]) -> None:
```